### DISCUSSION TITLE
Counting / slicing a 2.0 Result

### DISCUSSION THREAD

**Discussion by lafrech** [sqlalchemy/sqlalchemy#6334] [Usage Questions] [answered]
The 1.0 API allows a method to return a `Query` object that can be used as a cursor.

This is very nice because it basically allows a view function in my REST APIs to just return that cursor while the framework takes care of the rest (no pun intended...).

More in [the framework docs](https://flask-smorest.readthedocs.io/en/latest/pagination.html) for the curious reader but a typical route can be just that one-liner:

```py
@blp.route('/')
class Pets(MethodView):

    @blp.response(200, PetSchema(many=True))
    @blp.paginate(SQLCursorPage)
    def get(self):
        return db.session.query(Pets)
```

The 2.0 API version would look like this:

```py
@blp.route('/')
class Pets(MethodView):

    @blp.response(200, PetSchema(many=True))
    @blp.paginate(SQLCursorPage)
    def get(self):
        return db.session.execute(sqla.select(Pets))
```

except I miss two features in the result cursor that is returned.

### 1/ It can't count itself

     AttributeError: 'ChunkedIteratorResult' object has no attribute 'count'

A bit of searching led me to this [SO question](https://stackoverflow.com/q/65754023/4653485). It says that indeed the new cursor doesn't have the count feature, so it must be done explicitly. While the motivations behind this are understandable, it breaks the use case above.

The way to go would be to use a subquery

```py
        query = sqla.select(Pets)
        count = db.session.execute(
                sqla.select(sqla.func.count()).select_from(query.subquery())
            ).scalars().one()
        items = db.session.execute(query)
```

This could fit my need if `query` was somehow accessible from the `Result` object. While building `Result`, `execute` could store a reference to the query object for information / further access.

Then the view function would return the result of the executed query and the framework could compute the count outside of the view function.

### 2/ It can't be sliced

    TypeError: 'ChunkedIteratorResult' object is not subscriptable

This is the second issue with pagination. Looks like here again, when the result set is computed, it is already too late to slice because that would mean modifying a query (adding offset / limit) while it is already executed.

-------------------------------------------------

While writing this question, I managed to get my example working by returning the query rather than the cursor (the executed query):

```py
@blp.route('/')
class Pets(MethodView):

    @blp.response(200, PetSchema(many=True))
    @blp.paginate(SQLCursorPage)
    def get(self):
        return sqla.select(Pets)
```

and to perform the execution from my framework pagination class:

```py
class SQLCursorPage(flask_smorest.Page):
    """SQL cursor pager"""

    @property
    def items(self):
        return db.session.execute(
            self.query
            .offset(self.page_params.first_item)
            .limit(self.page_params.last_item + 1)
        ).scalars().all()

    @property
    def item_count(self):
        return db.session.execute(
            sqla.select(sqla.func.count())
            .select_from(self.query.subquery())
        ).scalars().one()
```

This works. What bothers me is that now my paginated route look different *because* it is paginated. I can't just remove the pagination decorator and get the whole list, for instance. This smells like bad design (from my application / framework, I mean) but I couldn't come up with a better one that would handle the 2.0 API.

I guess I answered my own question, but I'm posting it anyway in case it can help future readers or attract useful feedback. And also to mention that I miss the versatility of the `Query` object (while I'm totally open to the fact that this versatility might be the very reason it was deprecated).

To be totally transparent, I had a similar issue with the Python MongoDB driver, which stopped returning a countable cursor: https://jira.mongodb.org/browse/PYTHON-1724. It really looks like the original assumption in my framework's pagination feature (the fact that a cursor can be counted and sliced) was abusive.

Perhaps the count issue discussed in the SO question quoted above could be mentioned explicitly in the migration doc. But I won't complain, this doc is really good and has been quite helpful already.

**Comment by zzzeek** [comment, marked as answer]
> The 1.0 API allows a method to return a `Query` object that can be used as a cursor.

so it's actually not a cursor, which is why 2.0's API now makes it explicit at the point at which you get a "cursor", which is the session.execute() method.

> 
> This is very nice because it basically allows a view function in my REST APIs to just return that cursor while the framework takes care of the rest (no pun intended...).

This is fine, using 2.0 style you can result = session.execute(stmt) and that "result" is your "cursor".



> 
> More in [the framework docs](https://flask-smorest.readthedocs.io/en/latest/pagination.html) for the curious reader but a typical route can be just that one-liner:
> 
> ```python
> @blp.route('/')
> class Pets(MethodView):
> 
>     @blp.response(200, PetSchema(many=True))
>     @blp.paginate(SQLCursorPage)
>     def get(self):
>         return db.session.query(Pets)

ah you want a *paginated* "cursor", using the slices/index access and all that.  So yes, this is where everyone gets into trouble.      It's in fact extremely inefficient to paginate using LIMIT/OFFSET, which is what the slice feature has no alternative but to use.  If you wanted to paginate to a page 10000 rows in, the database has to scan all 10K rows to get there.  This is how applications that work great when you first write them and then for some reason perform terribly a year later get written :)

So in the new API we've necessarily moved away from making this not very scalable pattern implicit.



> ```
> 
> The 2.0 API version would look like this:
> 
> ```python
> @blp.route('/')
> class Pets(MethodView):
> 
>     @blp.response(200, PetSchema(many=True))
>     @blp.paginate(SQLCursorPage)
>     def get(self):
>         return db.session.execute(sqla.select(Pets))
> ```
> 
> except I miss two features in the result cursor that is returned.
> ### 1/ It can't count itself
> 
> ```
>  AttributeError: 'ChunkedIteratorResult' object has no attribute 'count'

Correct - use select(func.count()).select_from(Pets) for that.  You can write your own "cursor" like object that does this.   The query.count() feature produces inefficient queries as it has to wrap everything in a subquery, because it can't assume what it's selecting from.    It's better for people to write out the SQL they want exactly.

> ```
> 
> A bit of searching led me to this [SO question](https://stackoverflow.com/q/65754023/4653485). It says that indeed the new cursor doesn't have the count feature, so it must be done explicitly. While the motivations behind this are understandable, it breaks the use case above.

I disagree anything is "broken" here, query.count() still works and is not being removed from SQLAlchemy, it's just not the best way to do things.

> 
> The way to go would be to use a subquery
> 
> ```python
>         query = sqla.select(Pets)
>         count = db.session.execute(
>                 sqla.select(sqla.func.count()).select_from(query.subquery())
>             ).scalars().one()
>         items = db.session.execute(query)
> ```
> 
> This could fit my need if `query` was somehow accessible from the `Result` object. While building `Result`, `execute` could store a reference to the query object for information / further access.

I think a user-defined Pagination object here is the best approach.  SQLAlchemy itself is the library providing the SQL engine but we want to get out of the business of guessing and implicitly falling back to inefficient SQL patterns that don't scale.    These are things that should be considered by the developer.



> 
> Then the view function would return the result of the executed query and the framework could compute the count outside of the view function.
> ### 2/ It can't be sliced
> 
> ```
> TypeError: 'ChunkedIteratorResult' object is not subscriptable
> ```
> 
> This is the second issue with pagination. Looks like here again, when the result set is computed, it is already too late to slice because that would mean modifying a query (adding offset / limit) while it is already executed.
> 
> While writing this question, I managed to get my example working by returning the query rather than the cursor (the executed query):

pagination is actually hard, and there's a lot of ways to do it:

https://use-the-index-luke.com/no-offset

above, Markus explains that good pagination is unfortunately very specific to the table structure that's being SELECTed.   We also have a series of recipes here: https://github.com/sqlalchemy/sqlalchemy/wiki/RangeQuery-and-WindowedRangeQuery   

these are the more *correct* ways to do pagination, and they are not as simple as LIMIT/OFFSET unfortunately.   they aren't candidates to be generically stuffed into a slice operator, but of course a user-defined Pagination object that places specific constraints on the kind of table / object being paginated can make more fine-grained decisions on how to do this.

> This works. 

it "works" , but it wont "work" if you have 1M rows,m unfortunately, because OFFSET does not scale.


> What bothers me is that now my paginated route look different _because_ it is paginated. I can't just remove the pagination decorator and get the whole list, for instance. This smells like bad design (from my application / framework, I mean) but I couldn't come up with a better one that would handle the 2.0 API.
> 
> I guess I answered my own question, but I'm posting it anyway in case it can help future readers or attract useful feedback. And also to mention that I miss the versatility of the `Query` object (while I'm totally open to the fact that this versatility might be the very reason it was deprecated).
> 
> To be totally transparent, I had a similar issue with the Python MongoDB driver, which stopped returning a countable cursor: https://jira.mongodb.org/browse/PYTHON-1724. It really looks like the original assumption in my framework's pagination feature (the fact that a cursor can be counted and sliced) was abusive.

interesting they have a totally different architecture but the same issue.  pagination is hard, unfortunately.


> 
> Perhaps the count issue discussed in the SO question quoted above could be mentioned explicitly in the migration doc. But I won't complain, this doc is really good and has been quite helpful already.


+1

**Reply by lafrech** [reply]
Thank you @zzzeek for this detailed answer and for the instructive links.

I understand the rationale for the change and the will to make things explicit.

Good point about scalability.

Let me share a few thoughts here.

I see three classes of data/tables in (at least my) applications:

- Very high volume, where offset becomes problematic. I think my only real use case here is timeseries, and I don't use basic pagination here, rather selection by dates/timestamp.
- Low volume, for which I could probably get away without any pagination at all.
- Middle volume, where offset does the job well but the number of rows is high enough to justify the use of a pagination.

I won't put precise figures in front of that but arguably, the offset scaling issue comes for huge amounts of data (millions), while the use of a pagination to limit DB and network bandwidth can be interesting starting with lower amounts (hundreds).

This leaves room for the use of naive offset pagination, considering it is easy to implement in a generic fashion and it works for most tables. I generally have one huge (timeseries) table and many low size tables.

(I realize network bandwidth is another issue and you may fetch the whole result from the DB while still paginating the result but it's a pity and besides, when using an ORM, you may end up instantiating objects you're not using, so from a practical perspective I don't see a use case for that.)

-------------------------------------------------------------------------------------

In some cases, it is likely that the pagination I'm adding just gets in the way of the front-end, creating issues when rows are added/removed between two page fetches, as pointed out in the links, so if the row number is known to be low (small table or mandatory filter that makes the result set small) I should probably remove the pagination. In fact, I already saw frontend code fetching all pages to build the whole list which I consider a red flag for the reason mentioned above about the lost row and because it defeats the purpose of pagination.

For the cases where I still consider pagination useful, I should think of the actual use cases and perhaps consider keyset pagination. It sounds appealing but it is also an important change client side. More infinite scrolling oriented than table oriented. Perhaps frontends use tables precisely because backends used offset pagination first.

**At least those links made me realize that pagination doesn't come for free or as cheap as I thought and should be designed carefully depending on the use case.** Kind of like indexes.

![I've learned something today](https://64.media.tumblr.com/avatar_dba324a73724_128.pnj)


--------------------------------------------------------------------------------------

Back to my framework, I think the offset-based pager class will work in most cases, those cases where offset is just enough, so I'll probably be keeping that implementation in my app.

The framework also allows to paginate "manually" inside a view function, so the feature is not forced on the user. It is just a shortcut that assumes stuff. A note about the scaling issues could be added to the framework docs, though.

--------------------------------------------------------------------------------------

> I disagree anything is "broken" here, query.count() still works and is not being removed from SQLAlchemy, it's just not the best way to do things.

Oh, right. I was in the process of rewriting all queries as per the table at https://docs.sqlalchemy.org/en/14/changelog/migration_20.html#migration-orm-usage to remove the deprecation warnings and I thought I had to change those queries as well. I was wrong. The lines generating a warning are the "query(User).get" ones. The docs do mention that "there is no plan to actually remove the Session.query() API itself".
